src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,395 @@
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+symbolt &c_new_tmp_symbol(
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  const typet &type,
+  const std::string &prefix)
+{
+  return get_fresh_aux_symbol(
+    type,
+    "",
+    prefix,
+    loc,
+    ID_C,
+    symbol_table);
+}
+
+/*******************************************************************\
+
+Function: c_declare_new_tmp_symbol
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+symbolt &c_declare_new_tmp_symbol(
+  code_blockt &init_code,
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  const typet &type,
+  const std::string &prefix)
+{
+  symbolt &tmp_symbol=c_new_tmp_symbol(
+    symbol_table,
+    loc,
+    type,
+    prefix);
+
+  code_declt decl;
+  decl.add_source_location()=loc;
+  decl.symbol()=tmp_symbol.symbol_expr();
+  init_code.add(decl);
+
+  return tmp_symbol;
+}
+
+/*******************************************************************\
+
+Function: c_get_nondet_bool
+
+ Inputs: Desired type (C_bool or plain bool)
+
+ Outputs: nondet expr of that type
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt c_get_nondet_bool(const typet &type)
+{
+  // We force this to 0 and 1 and won't consider
+  // other values.
+  return typecast_exprt(side_effect_expr_nondett(bool_typet()), type);
+}
+
+/*******************************************************************\
+
+Function: c_nondet_symbol_factory
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt c_nondet_symbol_factory(
+  const code_typet::parametert &p,
+  const int param_number,
+  const typet &type,
+  code_blockt &init_code,
+  bool allow_null,
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  message_handlert &message_handler,
+  input_recordert &input_recorder)
+{
+  irep_idt base_name=p.get_base_name().empty()?
+    ("argument#"+std::to_string(param_number)):p.get_base_name();
+  irep_idt identifier=id2string(goto_functionst::entry_point())+
+    "::"+id2string(base_name);
+
+  {
+    auxiliary_symbolt new_symbol;
+    new_symbol.mode=ID_C;
+    new_symbol.is_static_lifetime=false;
+    new_symbol.name=identifier;
+    new_symbol.base_name=base_name;
+    new_symbol.type=type;
+
+    symbol_table.move(new_symbol);
+  }
+
+  symbol_exprt symbol_expr(identifier, type);
+
+  code_declt decl;
+  decl.symbol()=symbol_expr;
+
+  init_code.add(decl);
+
+  if(type.id()==ID_c_bool)
+  {
+    code_assignt assign(decl.symbol(), c_get_nondet_bool(type));
+    init_code.move_to_operands(assign);
+  }
+  else if(type.id()==ID_pointer)
+  {
+    symbol_factoryt state(
+      init_code,
+      !allow_null,
+      symbol_table,
+      message_handler,
+      input_recorder);
+    state.gen_nondet_init(symbol_expr, loc);
+  }
```

Comment:
I copied the java code. The pointer case seems to cover arrays and structs as well, though this patch doesn't properly deal with them.